acq4/devices/Scientifica/scientifica.py

```diff
@@ -397,51 +403,64 @@ def zeroZ(self):
+        return self._autoZero(axis=1)
 
     def autoZZero(self):
-        self._autoZero(axis=2)
-
-    def _autoZero(self, axis: int | None = None):
-        self.autoZeroBtn.processing("Moving...")
-        self.autoXZeroBtn.processing("Moving...")
-        self.autoYZeroBtn.processing("Moving...")
-        self.autoZZeroBtn.processing("Moving...")
-        self._autoZeroFuture = self._doAutoZero(axis)
-        if axis is None:
-            finisher = self.autoZeroFinished
-        else:
-            finisher = getattr(self, f"auto{'XYZ'[axis]}ZeroFinished")
-        self._autoZeroFuture.sigFinished.connect(finisher)
-        if self._autoZeroFuture.isDone():  # in case of instant completion
-            finisher()
+        self.sigBusyMoving.emit(self.autoZZeroBtn)
+        return self._autoZero(axis=2)
 
     @future_wrap
-    def _doAutoZero(self, axis: int = None, _future: Future = None):
+    def _autoZero(self, axis: int = None, _future: Future = None) -> None:
         self._savedLimits = self.dev.getLimits()
         try:
+            # confirm with user that movement is safe
+            alert = Qt.QMessageBox()
+            alert.setText("This will move the stage to its limit and set that as the zero position. Ready?")
+            alert.setInformativeText("Please ensure that the device is not obstructed.")
+            alert.setStandardButtons(Qt.QMessageBox.Ok | Qt.QMessageBox.Cancel)
+            if alert.exec_() != Qt.QMessageBox.Ok:
+                raise _future.StopRequested()
+
             self.dev.setLimits(None, None, None)
             pos = self.dev.globalPosition()
-            far_away = [-1e27, -1e27, 1e27]
+            dest = pos[:]
+            far_away = [-1e10, -1e10, 1e10]
             if axis is None:
-                pos = far_away
+                dest = far_away
             else:
-                pos[axis] = far_away[axis]
-            print(f"moving to {pos}")
-            self.dev.moveToGlobal(pos, "fast")
+                dest[axis] = far_away[axis]
+            self.dev.moveToGlobal(dest, "fast")
             _future.sleep(1)
             while self.dev.dev.isMoving():
                 _future.sleep(0.1)
             self.dev.stop()
-            self.dev.dev.zeroPosition('XYZ'[axis])
+            before = self.dev.globalPosition()
+            if axis is None:
+                self.dev.dev.zeroPosition()
+            else:
+                self.dev.dev.zeroPosition('XYZ'[axis])
+            self.dev.getPosition(refresh=True)
+            after = self.dev.globalPosition()
+            diff = np.array(after) - np.array(before)
+            logMsg(f"Auto-zeroed {self.dev.name()} by {diff}")
+            _future.waitFor(self.dev.moveToGlobal(pos + diff, "fast"))
+            dist = np.linalg.norm(diff)
+            if dist > 1e-2:
+                raise HelpfulException(
+                    f"Zeroing {self.dev.name()} moved by {siFormat(dist, suffix='m')}: check for accuracy.")
```

Comment:
I guessed about your intent here. Why does the user care?

Reply:
Large values (let's say more than 50 μm) indicate the axis has slipped significantly, which explains the cause of motion planning issues (particularly collisions) and might also tell us which rigs need to be repaired if they are consistently slipping more than others.